lib/sycamore/sycamore/transforms/summarize.py

```diff
@@ -61,133 +73,324 @@ def __init__(self, llm: LLM, element_operator: Optional[Callable[[Element], bool
 
+    Args:
+        llm: LLM to use for summarization
+        question: Optional question to use as context for the summarization. If set, the llm will
+            attempt to answer the question with the data provided
+        prompt: Prompt to use for each summarization. Caution: The default (MaxTokensHeirarchicalSummarizerPrompt)
+            has some fairly complicated logic encoded in it to make the tree construction work correctly.
+        fields: List of fields to include in each element's representation in the prompt. Specify
+            with dotted notation (e.g. properties.title), or use "*" to capture everything. If None,
+            will include no fields.
+        max_tokens: token limit for each summarization. Default is 10k (default tokenizer is by character).
+        tokenizer: tokenizer to use when computing how many tokens a prompt will take. Default is
+            CharacterTokenizer
+        rounds: number of rounds of heirarchical summarization to perform. The number of elements that can be
+            included in the summary is O(e^rounds), so rounds can be small. Default is 4.
+    """
 
-class QuestionAnsweringSummarizer:
-    def __init__(self, llm: LLM, question: str):
+    def __init__(
+        self,
+        llm: LLM,
+        question: Optional[str] = None,
+        data_description: Optional[str] = None,
+        prompt: SycamorePrompt = MaxTokensHeirarchyPrompt,
+        fields: Union[None, Literal["*"], list[str]] = None,
+        max_tokens: int = 10 * 1000,
+        tokenizer: Tokenizer = CharacterTokenizer(),
+        rounds: int = 4,
+    ):
         self.llm = llm
+        self.prompt = prompt.set(**self.get_const_vars())
+        self.fields = fields
         self.question = question
-
-    def __call__(self, text: str) -> str:
-        messages = SummarizeDataMessagesPrompt(question=self.question, text=text).as_messages()
-        prompt_kwargs = {"messages": messages}
-
-        t0 = time.time()
-        # call to LLM
-        summary = self.llm.generate_old(prompt_kwargs=prompt_kwargs, llm_kwargs={"temperature": 0})
-        t1 = time.time()
-        logging.info(f"Summarizer took {t1 - t0} seconds to generate summary.")
-
-        return summary
-
-
-def collapse(text: str, tokens_per_chunk: int, tokenizer: Tokenizer, summarizer_fn: Callable[[str], str]) -> str:
+        self.data_description = data_description
+        self.max_tokens = max_tokens
+        self.tokenizer = tokenizer
+        self.rounds = 4
+
+    @staticmethod
+    def get_const_vars() -> dict[str, str]:
+        return {
+            "skip_me_key": "_skip_me",
+            "batch_key": "_batch",
+            "round_key": "_round",
+            "intermediate_summary_key": "_summary",
+        }
+
+    def prep_batches(self, doc: Document, round: int = 0) -> Document:
+        vars = self.get_const_vars()
+        for i, elt in enumerate(doc.elements):
+            elt.properties[vars["round_key"]] = round
```

Comment:
I think you can simplify this by calculating a list of batches, but that would require changes to the way llmmapelements works, so lets chat.  Without that, this will need a bunch of explanation.